examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp

```diff
@@ -234,6 +234,201 @@ Status EnergyEvseDelegate::StartDiagnostics()
+            ChipLogError(AppServer, "DayOfWeekForSequence has a bit set which has already been set in another entry.");
+            return Status::ConstraintError;
+        }
+        dayOfWeekBitmap |= bitmask; // add this day Of week to the previously seen days
+
+        auto iterInner   = entry.chargingTargets.begin();
+        uint8_t innerIdx = 0;
+        while (iterInner.Next())
+        {
+            auto & targetStruct          = iterInner.GetValue();
+            uint16_t minutesPastMidnight = targetStruct.targetTimeMinutesPastMidnight;
+            ChipLogProgress(AppServer, "[%d] MinutesPastMidnight : %d", innerIdx,
+                            static_cast<short unsigned int>(minutesPastMidnight));
+
+            if (minutesPastMidnight > 1439)
+            {
+                ChipLogError(AppServer, "MinutesPastMidnight has invalid value (%d)", static_cast<int>(minutesPastMidnight));
+                return Status::ConstraintError;
+            }
+
+            uint8_t targetSoC;
+            int64_t addedEnergy;
+
+            /* Check that we have at least an AddedEnergy or TargetSoC value */
+            if (!(targetStruct.targetSoC.HasValue()) && !(targetStruct.addedEnergy.HasValue()))
+            {
+                ChipLogError(AppServer, "Must have one of AddedEnergy or TargetSoC");
+                return Status::ConstraintError;
+            }
+
+            if (targetStruct.targetSoC.HasValue())
+            {
+                targetSoC = targetStruct.targetSoC.Value();
+                ChipLogProgress(AppServer, "[%d] TargetSoC           : %d", innerIdx, static_cast<int>(targetSoC));
+
+                if (targetSoC > 100)
+                {
+                    ChipLogError(AppServer, "TargetSoC has invalid value (%d)", static_cast<int>(targetSoC));
+                    return Status::ConstraintError;
+                }
+                /* If SoC feature is not supported check that the value is 100% */
+                if (targetSoC != 100) // TODO work out how to check feature support in Delegate
+                {
+                    /* The only value we allow is 100 */
+                    ChipLogError(AppServer, "TargetSoC has can only be 100%% if SOC feature is not supported");
+                    return Status::ConstraintError;
+                }
+            }
+            if (targetStruct.addedEnergy.HasValue())
+            {
+                addedEnergy = targetStruct.addedEnergy.Value();
+                ChipLogProgress(AppServer, "[%d] AddedEnergy         : %ld", innerIdx, static_cast<signed long int>(addedEnergy));
+                if (addedEnergy < 0)
+                {
+                    ChipLogError(AppServer, "AddedEnergy has invalid value (%ld)", static_cast<signed long int>(addedEnergy));
+                    return Status::ConstraintError;
+                }
+            }
+            innerIdx++;
+        }
+    }
+
+    return Status::Success;
```

Comment:
This is missing checks that the TLV was actually valid, no?  As in, nothing checks for errors on the iterators.

This is the big drawback of passing the raw iterators to application code.  If we moved this validation into the cluster, that would at least be handled consistently there, leaving fewer pitfalls for app authors who may not realize all the checks they need to be doing.

Reply:
I have moved this validation method into the cluster (src/app/clusters/energy-evse-server/energy-evse-server.cpp) and added the TLV checks.